In [2]:
!pip install pyLDAvis==3.4.1

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 10.0 MB/s eta 0:00:00


In [3]:
!pip install -U git+https://github.com/stephenhky/PyShortTextCategorization

  Cloning https://github.com/stephenhky/PyShortTextCategorization to /tmp/pip-req-build-02bp203t
  Running command git clone --filter=blob:none --quiet https://github.com/stephenhky/PyShortTextCategorization /tmp/pip-req-build-02bp203t
  Resolved https://github.com/stephenhky/PyShortTextCategorization to commit c0cedef2fbda5576614f1740da35bf9cd0c6d57d
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 177.4/177.4 kB 2.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 20.5 MB/s eta 0:00:00
  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-

### **Preprocessing**

In [1]:
# Uncomment this code block if your data is in google drive
# from google.colab import drive
# drive.mount('/content/drive', force_remount=True)
# %cd /content/drive/My Drive/ColabData

Mounted at /content/drive
/content/drive/My Drive/ColabData


**Loading the files**

In [ ]:
# Importing the files
from nltk.corpus import PlaintextCorpusReader
# corpus_root = '/content/drive/MyDrive/ColabData/ceo-letters-shareholders'# Modify for your own google drive
# Uncomment the above variable corupus_root if running from Google Drive and
# Comment below 
corpus_root = './data/ceo-letters-shareholders/.'

In [4]:
filelists = PlaintextCorpusReader(corpus_root, '.*',encoding='latin-1')
file_names = filelists.fileids()
file_names #This will contain list of filenames

['3M_Roman_2020.txt',
 'ABM Industries_Salmirs_2022.txt',
 'ADM_Luciano_2021.txt',
 'AECOM_Rudd_2021.txt',
 'AES_Gluski_2020.txt',
 'AGCO_Hansotia_2021.txt',
 'AIG_Zaffino_2020.txt',
 'APA_Christmann_2021.txt',
 'ATT_Stankey_2021.txt',
 'AbbVie_Gonzalez_2021.txt',
 'Abbott Laboratories_Ford_2022.txt',
 'Activision Blizzard_Kotick_2020.txt',
 'Adobe_Narayen_2022.txt',
 'Advance Auto Parts_Greco_2021.txt',
 'Advanced Micro Devices_Su_2022.txt',
 'Aflac_Amos_2021.txt',
 'Air Products Chemicals_Ghasemi_2021.txt',
 'Alaska Air Group_Minicucci_2022.txt',
 'Albertsons_Sankaran_2020.txt',
 'Alcoa_Harvey_2021.txt',
 'AllState_Wislon_2021.txt',
 'Alleghany_Brandon_2022.txt',
 'Alliance Data Systems Bread Financial_Andretta_2021.txt',
 'Ally Financial_Brown_2021.txt',
 'Alphabet_Pichai_2020.txt',
 'Altria_Gifford_2021.txt',
 'Amazon_Bezos_2020.txt',
 'Ameren_Lyons_2022.txt',
 'American Axle Manufacturing_Dauch_2021.txt',
 'American Express_Squeri_2021.txt',
 'American Tower_Bartlett_2021.txt',
 '

In [5]:
#Loading the contents of the letters and storing them in a list
content = []
for fileid in filelists.fileids():
  content.append(filelists.raw(fileid))

In [6]:
len(content)

386

In [7]:
#Display a sample letter
content[1]

'Dear Stockholders, 2021 was an exceptional year for ABM, highlighted by strong financial performance and significant progress on multiple strategic fronts. Throughout the year, our global team continued to execute at the highest level, successfully navigating through a dynamic environment with agility, resiliency, and dedication to serve our clients, while at the same time, adapting to ever-changing COVID-19 protocols and advancing our brand and competitive positioning in our served markets. During 2021 we generated revenue of $6.2 billion, up 4% from 2020. Income from continuing operations (and net income) was $126.3 million, representing a significant increase over the prior year, while adjusted income from continuing operations* increased 49% to $243.3 million. Adjusted EBITDA* margin expanded to a record 7.3%, up 130 basis points from 2020, and significantly higher than our pre-COVID-19 pandemic level. 2021 also marked 56th consecutive year in which weâ\x80\x99ve paid a dividend, 

In [8]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
# from nltk.tokenize import sent_tokenize, word_tokenize
# sentences = [sent_tokenize(i) for i in content]
# sentences = [sentence for doc in sentences for sentence in doc]

In [9]:
#sentences[:1]

**Preprocessing and creating Document Term Matrix**

In [10]:
from shorttext.utils import standard_text_preprocessor_1, DocumentTermMatrix
preprocessor = standard_text_preprocessor_1()
corpus = [preprocessor(article).split(' ') for article in content]

We are using shortext librabry for text preprocessing each document in the corpus by splitting them into words and we are storing this as a list of words in variable (corpus).

In [11]:
#Display a sample letter after tokeniztion
content[1]

'Dear Stockholders, 2021 was an exceptional year for ABM, highlighted by strong financial performance and significant progress on multiple strategic fronts. Throughout the year, our global team continued to execute at the highest level, successfully navigating through a dynamic environment with agility, resiliency, and dedication to serve our clients, while at the same time, adapting to ever-changing COVID-19 protocols and advancing our brand and competitive positioning in our served markets. During 2021 we generated revenue of $6.2 billion, up 4% from 2020. Income from continuing operations (and net income) was $126.3 million, representing a significant increase over the prior year, while adjusted income from continuing operations* increased 49% to $243.3 million. Adjusted EBITDA* margin expanded to a record 7.3%, up 130 basis points from 2020, and significantly higher than our pre-COVID-19 pandemic level. 2021 also marked 56th consecutive year in which weâ\x80\x99ve paid a dividend, 

In [12]:
#Importing all the text into a dataframe using pandas
import pandas as pd
df = pd.DataFrame(content, columns = ['letters'])

In [13]:
df.head()

letters
0  Dear shareholders,\n\nAs the COVID-19 pandemic...
1  Dear Stockholders, 2021 was an exceptional yea...
2  Dear Stockholders, \n\nWhen I last reported to...
3  Dear Stockholders:\n\nAs we begin the new year...
4  When 2020 began, no one predicted that we woul...

In [14]:
df.shape

(386, 1)

In [15]:
#Appending filenames to the respective letters
df = df.assign(title = file_names)
df.head()

letters  \
0  Dear shareholders,\n\nAs the COVID-19 pandemic...   
1  Dear Stockholders, 2021 was an exceptional yea...   
2  Dear Stockholders, \n\nWhen I last reported to...   
3  Dear Stockholders:\n\nAs we begin the new year...   
4  When 2020 began, no one predicted that we woul...   

                             title  
0                3M_Roman_2020.txt  
1  ABM Industries_Salmirs_2022.txt  
2             ADM_Luciano_2021.txt  
3              AECOM_Rudd_2021.txt  
4              AES_Gluski_2020.txt

In [16]:
#Adding Index column
df['index'] = df.index

In [ ]:
df.head()

letters  \
0  Dear shareholders,\n\nAs the COVID-19 pandemic...   
1  Dear Stockholders, 2021 was an exceptional yea...   
2  Dear Stockholders, \n\nWhen I last reported to...   
3  Dear Stockholders:\n\nAs we begin the new year...   
4  When 2020 began, no one predicted that we woul...   

                             title  index  
0                3M_Roman_2020.txt      0  
1  ABM Industries_Salmirs_2022.txt      1  
2             ADM_Luciano_2021.txt      2  
3              AECOM_Rudd_2021.txt      3  
4              AES_Gluski_2020.txt      4

In [17]:
#View Row with Index=0
data_sample = df[df['index'] == 0]
data_sample

letters              title  index
0  Dear shareholders,\n\nAs the COVID-19 pandemic...  3M_Roman_2020.txt      0

In [18]:
doc_sample = data_sample.values[0][0]
doc_sample

'Dear shareholders,\n\nAs the COVID-19 pandemic worsened in March 2020, 3Mâ\x80\x99s respirator plant in Valley, Nebraska issued a call for help: it needed two dozen 3M engineers and technicians to work on additional manufacturing lines that were already running around the clock. By the next morning, over a hundred 3Mers from across the Midwest had volunteered â\x80\x94 at a momentâ\x80\x99s notice â\x80\x94 to pack up, temporarily relocate and sacrifice time away from family to make a difference. It is inspiring, but it is not surprising. This is 3M. This is who we are and what we do. I have been with this company for more than three decades. 2020 was a year like no other, and Iâ\x80\x99ve never been prouder of how 3Mers everywhere stepped up to help the world during a unique moment in history. Our response throughout COVID-19 has been guided by our purpose, and shaped by three priorities: protecting our employees, fighting the pandemic from every angle and delivering for our customer

In [19]:
#Perform preprocessing
import gensim
from gensim.utils import simple_preprocess
from gensim.parsing.preprocessing import STOPWORDS
from nltk.stem import WordNetLemmatizer, SnowballStemmer
from nltk.stem.porter import *
import numpy as np

In [23]:
from nltk.stem import*
stemmer = PorterStemmer()

Defining function to perform lemmatization and stemming preprocessing steps on the data set. Without contextual information the lemmatiser’s not able to distinguish between nouns and verbs. In this case, "pos = 'v'" means treat the words as verbs. We are performing the following:
*   Tokenization
*   Removing stopwords
*   Filter out tokens with length less than or equal to 3
*   Return preprocessed text as list of tokens





In [20]:
def lemmatize_stemming(text):
    return stemmer.stem(WordNetLemmatizer().lemmatize(text, pos='v'))

def preprocess(text):
    result = []
    for token in gensim.utils.simple_preprocess(text):
        if token not in gensim.parsing.preprocessing.STOPWORDS and len(token) > 3:
            result.append(lemmatize_stemming(token))
    return result

In [21]:
#remove na values from the column 'letters'
data_text = df.dropna(subset=['letters'])

In [24]:
#select a document to preview after preprocessing
import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')

words = []
for word in doc_sample.split(' '):
    words.append(word)
print(words)
print('\n\n tokenized and lemmatized document: ')
print(preprocess(doc_sample))

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


['Dear', 'shareholders,\n\nAs', 'the', 'COVID-19', 'pandemic', 'worsened', 'in', 'March', '2020,', '3Mâ\x80\x99s', 'respirator', 'plant', 'in', 'Valley,', 'Nebraska', 'issued', 'a', 'call', 'for', 'help:', 'it', 'needed', 'two', 'dozen', '3M', 'engineers', 'and', 'technicians', 'to', 'work', 'on', 'additional', 'manufacturing', 'lines', 'that', 'were', 'already', 'running', 'around', 'the', 'clock.', 'By', 'the', 'next', 'morning,', 'over', 'a', 'hundred', '3Mers', 'from', 'across', 'the', 'Midwest', 'had', 'volunteered', 'â\x80\x94', 'at', 'a', 'momentâ\x80\x99s', 'notice', 'â\x80\x94', 'to', 'pack', 'up,', 'temporarily', 'relocate', 'and', 'sacrifice', 'time', 'away', 'from', 'family', 'to', 'make', 'a', 'difference.', 'It', 'is', 'inspiring,', 'but', 'it', 'is', 'not', 'surprising.', 'This', 'is', '3M.', 'This', 'is', 'who', 'we', 'are', 'and', 'what', 'we', 'do.', 'I', 'have', 'been', 'with', 'this', 'company', 'for', 'more', 'than', 'three', 'decades.', '2020', 'was', 'a', 'year',

In [25]:
#preprocess'TEXT', save the results as 'processed_docs'
processed_docs = data_text['letters'].map(preprocess)
processed_docs[:10]

0    [dear, sharehold, covid, pandem, worsen, march...
1    [dear, stockhold, except, year, highlight, str...
2    [dear, stockhold, report, year, close, monitor...
3    [dear, stockhold, begin, year, want, especi, t...
4    [begin, predict, face, biggest, public, health...
5    [year, theâ, world, wildlif, foundat, estim, f...
6    [dear, sharehold, honor, write, aigâ, presid, ...
7    [fellow, sharehold, challeng, experi, persist,...
8    [sharehold, employe, face, share, challeng, hi...
9    [dear, abbvi, sharehold, compani, deliv, outst...
Name: letters, dtype: object

Create a dictionary from ‘processed_docs’ containing the number of times a word appears in the document set

In [26]:
dictionary = gensim.corpora.Dictionary(processed_docs)

Filter out tokens that appear in less than 15 documents (absolute number) or more than 0.5 documents (fraction of total corpus size, not absolute number).After the above two steps, keep only the first 100000 most frequent tokens.
For each document we create a dictionary reporting how many
words and how many times those words appear. Save this to ‘bow_corpus’, then check our selected document earlier.

In [27]:
dictionary.filter_extremes(no_below=15, no_above=0.5, keep_n=100000)

In [28]:
bow_corpus = [dictionary.doc2bow(doc) for doc in processed_docs]
bow_corpus[0]

[(0, 4),
 (1, 1),
 (2, 1),
 (3, 2),
 (4, 1),
 (5, 2),
 (6, 1),
 (7, 1),
 (8, 1),
 (9, 1),
 (10, 2),
 (11, 1),
 (12, 1),
 (13, 1),
 (14, 1),
 (15, 1),
 (16, 1),
 (17, 1),
 (18, 1),
 (19, 1),
 (20, 1),
 (21, 1),
 (22, 1),
 (23, 1),
 (24, 1),
 (25, 1),
 (26, 1),
 (27, 1),
 (28, 3),
 (29, 1),
 (30, 3),
 (31, 3),
 (32, 3),
 (33, 1),
 (34, 1),
 (35, 1),
 (36, 1),
 (37, 1),
 (38, 1),
 (39, 2),
 (40, 1),
 (41, 1),
 (42, 1),
 (43, 1),
 (44, 1),
 (45, 1),
 (46, 1),
 (47, 1),
 (48, 1),
 (49, 1),
 (50, 1),
 (51, 2),
 (52, 1),
 (53, 1),
 (54, 2),
 (55, 1),
 (56, 1),
 (57, 1),
 (58, 1),
 (59, 1),
 (60, 3),
 (61, 2),
 (62, 1),
 (63, 1),
 (64, 2),
 (65, 1),
 (66, 1),
 (67, 4),
 (68, 2),
 (69, 1),
 (70, 1),
 (71, 1),
 (72, 1),
 (73, 1),
 (74, 1),
 (75, 1),
 (76, 3),
 (77, 3),
 (78, 1),
 (79, 1),
 (80, 1),
 (81, 1),
 (82, 1),
 (83, 2),
 (84, 2),
 (85, 1),
 (86, 1),
 (87, 4),
 (88, 1),
 (89, 1),
 (90, 1),
 (91, 2),
 (92, 3),
 (93, 2),
 (94, 2),
 (95, 1),
 (96, 2),
 (97, 1),
 (98, 1),
 (99, 1),
 (100, 3),

In [29]:
#viewing the bow_doc for sample 0 in dictionary
bow_doc_0 = bow_corpus[0]
for i in range(len(bow_doc_0)):
    print("Word {} (\"{}\") appears {} time.".format(bow_doc_0[i][0], dictionary[bow_doc_0[i][0]], bow_doc_0[i][1]))

Word 0 ("abil") appears 4 time.
Word 1 ("abl") appears 1 time.
Word 2 ("account") appears 1 time.
Word 3 ("activ") appears 2 time.
Word 4 ("adapt") appears 1 time.
Word 5 ("adjust") appears 2 time.
Word 6 ("affect") appears 1 time.
Word 7 ("aggress") appears 1 time.
Word 8 ("agil") appears 1 time.
Word 9 ("alloc") appears 1 time.
Word 10 ("allow") appears 2 time.
Word 11 ("ambiti") appears 1 time.
Word 12 ("america") appears 1 time.
Word 13 ("analyt") appears 1 time.
Word 14 ("answer") appears 1 time.
Word 15 ("anticip") appears 1 time.
Word 16 ("appli") appears 1 time.
Word 17 ("approxim") appears 1 time.
Word 18 ("asia") appears 1 time.
Word 19 ("attract") appears 1 time.
Word 20 ("automot") appears 1 time.
Word 21 ("avoid") appears 1 time.
Word 22 ("away") appears 1 time.
Word 23 ("balanc") appears 1 time.
Word 24 ("basi") appears 1 time.
Word 25 ("batteri") appears 1 time.
Word 26 ("break") appears 1 time.
Word 27 ("broadli") appears 1 time.
Word 28 ("capac") appears 3 time.
Word 2

### **Topic Modelling Using LDA**

In [30]:
#run LDA using bag of words
#Because LDA uses randomness in both training and inference steps, setting a random state is very important to control the randomness to make answers repeatable.
lda_model = gensim.models.LdaMulticore(bow_corpus, num_topics=10, id2word=dictionary, passes=2, workers=2,random_state=4)

/usr/lib/python3.10/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = os.fork()


In [31]:
#for each topic, we will explore the words occuring in that topic and its relative weight.

for idx, topic in lda_model.print_topics(-1): # print all the topics. (-1) means all the topics.
    print('Topic: {} \nWords: {}'.format(idx, topic))

Topic: 0 
Words: 0.008*"store" + 0.007*"brand" + 0.007*"home" + 0.006*"platform" + 0.005*"consum" + 0.005*"fiscal" + 0.005*"associ" + 0.004*"transform" + 0.004*"equiti" + 0.004*"retail"
Topic: 1 
Words: 0.010*"client" + 0.010*"energi" + 0.009*"percent" + 0.006*"bank" + 0.005*"equiti" + 0.005*"risk" + 0.004*"insur" + 0.004*"incom" + 0.004*"underwrit" + 0.004*"dividend"
Topic: 2 
Words: 0.007*"patient" + 0.006*"care" + 0.005*"data" + 0.005*"healthcar" + 0.005*"platform" + 0.004*"transform" + 0.004*"launch" + 0.004*"member" + 0.004*"network" + 0.004*"client"
Topic: 3 
Words: 0.010*"client" + 0.009*"proxi" + 0.008*"vote" + 0.006*"fiscal" + 0.006*"stockhold" + 0.005*"hold" + 0.004*"store" + 0.004*"account" + 0.004*"consum" + 0.004*"state"
Topic: 4 
Words: 0.013*"bank" + 0.011*"consum" + 0.008*"brand" + 0.005*"percent" + 0.005*"platform" + 0.005*"client" + 0.004*"food" + 0.004*"payment" + 0.004*"govern" + 0.004*"credit"
Topic: 5 
Words: 0.006*"power" + 0.006*"transform" + 0.005*"care" + 0.00

In [32]:
#Viewing the topic coverage for sample document (#0)
lda_model[bow_corpus[0]]

[(0, 0.016885389), (2, 0.3540674), (6, 0.27470955), (8, 0.35284814)]

In [33]:
#check the topic distribution for the Document# 0.
#"tup:-1tup[1]" is a parameter for the key lambda for the sorted function. It means sorting by the reverse order of the topic socre.
for index, score in sorted(lda_model[bow_corpus[0]], key=lambda tup: -1*tup[1]):
    print("\nScore: {}\t \nTopic: {}".format(score, lda_model.print_topic(index, 10))) #for each topic, print the top 10 words in that topic.


Score: 0.3704669177532196	 
Topic: 0.007*"patient" + 0.006*"care" + 0.005*"data" + 0.005*"healthcar" + 0.005*"platform" + 0.004*"transform" + 0.004*"launch" + 0.004*"member" + 0.004*"network" + 0.004*"client"

Score: 0.32660675048828125	 
Topic: 0.005*"adjust" + 0.005*"transform" + 0.005*"margin" + 0.005*"segment" + 0.005*"acquisit" + 0.004*"emiss" + 0.004*"data" + 0.004*"energi" + 0.004*"profit" + 0.004*"connect"

Score: 0.285773366689682	 
Topic: 0.014*"energi" + 0.006*"power" + 0.005*"vehicl" + 0.005*"carbon" + 0.005*"project" + 0.005*"emiss" + 0.004*"electr" + 0.004*"unit" + 0.004*"renew" + 0.004*"fiscal"

Score: 0.015663495287299156	 
Topic: 0.008*"store" + 0.007*"brand" + 0.007*"home" + 0.006*"platform" + 0.005*"consum" + 0.005*"fiscal" + 0.005*"associ" + 0.004*"transform" + 0.004*"equiti" + 0.004*"retail"


In [34]:
#check the topic distribution for the all Documents.
#"tup:-1tup[1]" is a parameter for the key lambda for the sorted function. It means sorting by the reverse order of the topic socre.
#For clearly viewing the topics for coverage I have stored the output in a dataframe (final_df)
final_df = pd.DataFrame()
documents = []
scores = []
topics = []
for i in range(len(bow_corpus)):
  for index, score in lda_model[bow_corpus[i]]:#, key=lambda tup: -1*tup[1]):
    documents.append(i)
    scores.append(score)
    topics.append(lda_model.print_topic(index, 10))
    print("\nDocument:{}".format(i))
    print("\nScore: {}\t \nTopic: {}".format(score, lda_model.print_topic(index, 10))) #for each topic, print the top 10 words in that topic.
final_df["Document"] = documents
final_df["Score"] = scores
final_df["Topic"] = topics

Streaming output truncated to the last 5000 lines.

Document:1

Score: 0.3166562616825104	 
Topic: 0.005*"adjust" + 0.005*"transform" + 0.005*"margin" + 0.005*"segment" + 0.005*"acquisit" + 0.004*"emiss" + 0.004*"data" + 0.004*"energi" + 0.004*"profit" + 0.004*"connect"

Document:2

Score: 0.1202269047498703	 
Topic: 0.007*"patient" + 0.006*"care" + 0.005*"data" + 0.005*"healthcar" + 0.005*"platform" + 0.004*"transform" + 0.004*"launch" + 0.004*"member" + 0.004*"network" + 0.004*"client"

Document:2

Score: 0.25668150186538696	 
Topic: 0.013*"bank" + 0.011*"consum" + 0.008*"brand" + 0.005*"percent" + 0.005*"platform" + 0.005*"client" + 0.004*"food" + 0.004*"payment" + 0.004*"govern" + 0.004*"credit"

Document:2

Score: 0.07505172491073608	 
Topic: 0.006*"power" + 0.006*"transform" + 0.005*"care" + 0.005*"energi" + 0.004*"great" + 0.004*"allow" + 0.004*"percent" + 0.004*"data" + 0.004*"purpos" + 0.004*"brand"

Document:2

Score: 0.35873010754585266	 
Topic: 0.014*"energi" + 0.006*"power

In [35]:
#Viewing the dataframe
final_df

Document     Score                                              Topic
0            0  0.011012  0.008*"store" + 0.007*"brand" + 0.007*"home" +...
1            0  0.370535  0.007*"patient" + 0.006*"care" + 0.005*"data" ...
2            0  0.283948  0.014*"energi" + 0.006*"power" + 0.005*"vehicl...
3            0  0.333015  0.005*"adjust" + 0.005*"transform" + 0.005*"ma...
4            1  0.041408  0.008*"store" + 0.007*"brand" + 0.007*"home" +...
...        ...       ...                                                ...
1001       383  0.172788  0.013*"bank" + 0.011*"consum" + 0.008*"brand" ...
1002       383  0.028299  0.014*"energi" + 0.006*"power" + 0.005*"vehicl...
1003       384  0.675757  0.007*"patient" + 0.006*"care" + 0.005*"data" ...
1004       384  0.318646  0.005*"adjust" + 0.005*"transform" + 0.005*"ma...
1005       385  0.997105  0.007*"patient" + 0.006*"care" + 0.005*"data" ...

[1006 rows x 3 columns]

In [36]:
df.head()

letters  \
0  Dear shareholders,\n\nAs the COVID-19 pandemic...   
1  Dear Stockholders, 2021 was an exceptional yea...   
2  Dear Stockholders, \n\nWhen I last reported to...   
3  Dear Stockholders:\n\nAs we begin the new year...   
4  When 2020 began, no one predicted that we woul...   

                             title  index  
0                3M_Roman_2020.txt      0  
1  ABM Industries_Salmirs_2022.txt      1  
2             ADM_Luciano_2021.txt      2  
3              AECOM_Rudd_2021.txt      3  
4              AES_Gluski_2020.txt      4

Understanding themes in documents based on topics generated by model

In [ ]:
pd.reset_option("max_colwidth")
df.loc[df['title'] == 'Microsoft_Nadella_2021.txt']

letters  \
239  Dear shareholders, colleagues, customers, and ...   

                          title  index  
239  Microsoft_Nadella_2021.txt    239

In [ ]:
pd.set_option("max_colwidth",None)
final_df.loc[final_df['Document']==239]

Document     Score  \
631       239  0.365792   
632       239  0.533283   
633       239  0.037950   
634       239  0.058796   

                                                                                                                                                                             Topic  
631       0.008*"store" + 0.007*"brand" + 0.007*"home" + 0.006*"platform" + 0.005*"consum" + 0.005*"fiscal" + 0.005*"associ" + 0.004*"transform" + 0.004*"equiti" + 0.004*"retail"  
632  0.007*"patient" + 0.006*"care" + 0.005*"data" + 0.005*"healthcar" + 0.005*"platform" + 0.004*"transform" + 0.004*"launch" + 0.004*"member" + 0.004*"network" + 0.004*"client"  
633            0.006*"power" + 0.006*"transform" + 0.005*"care" + 0.005*"energi" + 0.004*"great" + 0.004*"allow" + 0.004*"percent" + 0.004*"data" + 0.004*"purpos" + 0.004*"brand"  
634           0.015*"ourâ" + 0.012*"andâ" + 0.010*"energi" + 0.006*"clean" + 0.004*"carbon" + 0.004*"power" + 0.004*"climat" + 0.004*"strengthen" + 0.003*"packag" + 0.003*"emiss"

Topics Identified by LDA for the document *Microsoft_Nadella_2021.txt* :\
**Retail Transformation**: Discussions on transforming retail operations, focusing on stores, brands, and consumer platforms.\
**Healthcare Platform and Data**: Conversations about healthcare systems, patient care, and data management within healthcare platforms.\
**Energy and Power**: Talks about energy transformation, power generation, and the pursuit of energy efficiency and renewable sources.\
**Environmental Concerns**: Dialogues addressing environmental issues such as climate change, carbon emissions, and efforts to strengthen environmental regulations.

In [ ]:
pd.reset_option("max_colwidth")
df.loc[df['title'] == 'DXC Technology_Salvino_2021.txt']

letters  \
112  I hope you and your families are doing well.\n...   

                               title  index  
112  DXC Technology_Salvino_2021.txt    112

In [ ]:
pd.set_option("max_colwidth",None)
final_df.loc[final_df['Document']==112]

Document     Score  \
291       112  0.025560   
292       112  0.849281   
293       112  0.118956   

                                                                                                                                                                             Topic  
291  0.007*"patient" + 0.006*"care" + 0.005*"data" + 0.005*"healthcar" + 0.005*"platform" + 0.004*"transform" + 0.004*"launch" + 0.004*"member" + 0.004*"network" + 0.004*"client"  
292            0.016*"client" + 0.006*"asset" + 0.005*"incom" + 0.005*"brand" + 0.005*"acquisit" + 0.005*"adjust" + 0.004*"profit" + 0.004*"care" + 0.004*"consum" + 0.004*"price"  
293           0.015*"ourâ" + 0.012*"andâ" + 0.010*"energi" + 0.006*"clean" + 0.004*"carbon" + 0.004*"power" + 0.004*"climat" + 0.004*"strengthen" + 0.003*"packag" + 0.003*"emiss"

Topics Identified by LDA for the document *DXC Technology_Salvino_2021.txt* :\
**Healthcare Platform and Data**: Discussion revolves around healthcare topics such as patient care, data management, and healthcare platforms.\
**Client Asset Management**: Focuses on client-related discussions, particularly concerning asset management, brand management, and financial aspects like income and profit.\
**Environmental Concerns**: Addresses environmental issues such as energy, clean energy initiatives, carbon emissions, and efforts to strengthen environmental policies.

In [ ]:
pd.reset_option("max_colwidth")
df.loc[df['title'] == 'Walmart_McMillon_2021.txt']

letters  \
366  To Our Shareholders\n\n\n\nLooking back at 202...   

                         title  index  
366  Walmart_McMillon_2021.txt    366

In [ ]:
pd.set_option("max_colwidth",None)
final_df.loc[final_df['Document']==366]

Document     Score  \
975       366  0.438591   
976       366  0.553137   

                                                                                                                                                                        Topic  
975  0.008*"store" + 0.007*"brand" + 0.007*"home" + 0.006*"platform" + 0.005*"consum" + 0.005*"fiscal" + 0.005*"associ" + 0.004*"transform" + 0.004*"equiti" + 0.004*"retail"  
976       0.006*"power" + 0.006*"transform" + 0.005*"care" + 0.005*"energi" + 0.004*"great" + 0.004*"allow" + 0.004*"percent" + 0.004*"data" + 0.004*"purpos" + 0.004*"brand"

Topics Identified by LDA for the document *Walmart_McMillon_2021.txt* :\
**Retail Transformation**: Discussion focuses on transformations within the retail industry, including aspects like stores, brands, and consumer behavior.\
**Energy and Power**: Centers around discussions related to energy and power, encompassing transformation efforts and considerations about energy sources and usage.

In [ ]:
pd.reset_option("max_colwidth")
df.loc[df['title'] == 'Home Depot_Menear_2021.txt']

letters  \
182  Dear Shareholders: \n\nOperating in an Unprece...   

                          title  index  
182  Home Depot_Menear_2021.txt    182

In [ ]:
pd.set_option("max_colwidth",None)
final_df.loc[final_df['Document']==182]

Document     Score  \
463       182  0.996702   

                                                                                                                                                                        Topic  
463  0.008*"store" + 0.007*"brand" + 0.007*"home" + 0.006*"platform" + 0.005*"consum" + 0.005*"fiscal" + 0.005*"associ" + 0.004*"transform" + 0.004*"equiti" + 0.004*"retail"

Topics Identified by LDA for the document *Home Depot_Menear_2021.txt* :\
**Retail Transformation**: The document primarily discusses transformations within the retail sector, covering topics such as stores, brands, consumer behavior, and fiscal considerations.

In [ ]:
pd.reset_option("max_colwidth")
df.loc[df['title'] == 'Southwest Airlines_Kelly_2021.txt']

letters  \
320  To our Shareholders:\n\n It was just over a ye...   

                                 title  index  
320  Southwest Airlines_Kelly_2021.txt    320

In [ ]:
pd.set_option("max_colwidth",None)
final_df.loc[final_df['Document']==320]

Document     Score  \
861       320  0.057129   
862       320  0.902281   
863       320  0.021595   
864       320  0.017975   

                                                                                                                                                                             Topic  
861       0.008*"store" + 0.007*"brand" + 0.007*"home" + 0.006*"platform" + 0.005*"consum" + 0.005*"fiscal" + 0.005*"associ" + 0.004*"transform" + 0.004*"equiti" + 0.004*"retail"  
862        0.010*"client" + 0.010*"energi" + 0.009*"percent" + 0.006*"bank" + 0.005*"equiti" + 0.005*"risk" + 0.004*"insur" + 0.004*"incom" + 0.004*"underwrit" + 0.004*"dividend"  
863  0.007*"patient" + 0.006*"care" + 0.005*"data" + 0.005*"healthcar" + 0.005*"platform" + 0.004*"transform" + 0.004*"launch" + 0.004*"member" + 0.004*"network" + 0.004*"client"  
864         0.013*"bank" + 0.011*"consum" + 0.008*"brand" + 0.005*"percent" + 0.005*"platform" + 0.005*"client" + 0.004*"food" + 0.004*"payment" + 0.004*"govern" + 0.004*"credit"

Topics Identified by LDA for the document *Southwest Airlines_Kelly_2021.txt* :\
**Retail Focus**: This document emphasizes on retail-related topics such as stores, brands, and consumer behavior.\
**Financial and Energy**: Delves into financial services, particularly client management, alongside energy-related discussions.\
**Healthcare Emphasis**: Centers around healthcare topics, including patient care and healthcare platforms.\
**Banking Operations**: Primarily discusses banking operations, consumer behavior, and brand management.

In [ ]:
pd.reset_option("max_colwidth")
df.loc[df['title'] == 'United Airlines_Kirby_2021.txt']

letters  \
349  Dear Stockholder: \n\nAt the start of 2020, ou...   

                              title  index  
349  United Airlines_Kirby_2021.txt    349

In [ ]:
pd.set_option("max_colwidth",None)
final_df.loc[final_df['Document']==349]

Document     Score  \
939       349  0.995671   

                                                                                                                                                                    Topic  
939  0.015*"ourâ" + 0.012*"andâ" + 0.010*"energi" + 0.006*"clean" + 0.004*"carbon" + 0.004*"power" + 0.004*"climat" + 0.004*"strengthen" + 0.003*"packag" + 0.003*"emiss"

Topics Identified by LDA for the document *United Airlines_Kirby_2021.txt* :\
**Environmental Focus**: The document predominantly discusses environmental concerns, emphasizing topics such as clean energy initiatives, carbon emissions, and efforts to address climate change.

##**Coherence Score**

In [38]:
from gensim.models.coherencemodel import CoherenceModel
coherence_model_lda = CoherenceModel(model=lda_model, texts=processed_docs, dictionary=dictionary, coherence='c_v')
coherence_lda = coherence_model_lda.get_coherence()
print(f'Coherence Score: {coherence_lda}')

Coherence Score: 0.3127463082155243


The coherence score is a measure of the quality or coherence of the topics generated by the LDA model. Specifically, it quantifies how interpretable and coherent the topics are based on the relationships between words within each topic.\
A coherence score of **0.3127** suggests that the topics generated by the LDA model are moderately coherent and interpretable.

### **Intertopic Distance Mapping**


Intertopic distance mapping is a visualization technique used in topic modeling to represent the relationships between topics identified by the LDA model. This visualization helps to understand the similarities and differences between topics and also help us interprete the tone of each topic.

In [ ]:
import pyLDAvis
pyLDAvis.enable_notebook()
import pyLDAvis.gensim_models as gensimvis #This only works with pyLDAvis 3.4.0 on GoogleColab


In [ ]:
vis = gensimvis.prepare(lda_model, bow_corpus, dictionary)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/pyLDAvis/_prepare.py:244: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  by='saliency', ascending=False).head(R).drop('saliency', 1)
/usr/local/lib/python3.10/dist-packages/pandas/core/dtypes/cast.py:1641: DeprecationWarning: np.find_common_type is deprecated.  Please use `np.result_type` or `np.promote_types`.
See https://numpy.org/devdocs/release/1.25.0-notes.html and the docs for more information.  (Deprecated NumPy 1.25)
  return np.find_common_type(types, [])


In [ ]:
vis

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
0     -0.012629  0.018227       1        1  15.852169
2      0.000286  0.015535       2        1  15.295281
4     -0.015585 -0.001953       3        1  15.189995
6      0.028176 -0.006731       4        1  14.457922
7     -0.011630 -0.002355       5        1  12.483077
8      0.009791  0.008093       6        1   9.579412
1      0.000918 -0.029066       7        1   9.125253
3     -0.030847 -0.009866       8        1   4.198376
5      0.006434  0.010880       9        1   2.634599
9      0.025086 -0.002764      10        1   1.183914, topic_info=        Term        Freq       Total Category  logprob  loglift
80    energi  785.000000  785.000000  Default  30.0000  30.0000
284   client  897.000000  897.000000  Default  29.0000  29.0000
1332    bank  641.000000  641.000000  Default  28.0000  28.0000
35     clean  158.000000  158.000000  Default  27.0000  27.0000
532    store  443.000000  443.000000  Default  26.0000  26.0000
...      ...         ...         ...      ...      ...      ...
33     chain    4.841358  404.711069  Topic10  -5.9324   0.0104
353   profit    4.917719  457.863535  Topic10  -5.9167  -0.0974
284   client    5.068623  897.644875  Topic10  -5.8865  -0.7404
51      data    4.798862  529.731683  Topic10  -5.9412  -0.2676
111   govern    4.726458  444.518573  Topic10  -5.9564  -0.1075

[669 rows x 6 columns], token_table=      Topic      Freq  Term
term                       
0         1  0.158537  abil
0         2  0.212440  abil
0         3  0.187074  abil
0         4  0.076098  abil
0         5  0.126830  abil
...     ...       ...   ...
590       6  0.087298  zero
590       7  0.174596  zero
590       8  0.019400  zero
590       9  0.024249  zero
590      10  0.014550  zero

[4173 rows x 3 columns], R=30, lambda_step=0.01, plot_opts={'xlab': 'PC1', 'ylab': 'PC2'}, topic_order=[1, 3, 5, 7, 8, 9, 2, 4, 6, 10])

We can observe from above that Topic 1 has high relevance for words like store, brand, home, platform, consume indicating it could be more about retail and branding.